### DISCUSSION TITLE
Per-user security for realtime server

### DISCUSSION THREAD

**Discussion by sswatson** [supabase/supabase#1550] [Questions] [answered]
The docs say that the realtime server doesn't support per-user security yet. Does this mean that any tables with realtime enabled need to be public (for now)?

I'm having a hard time squaring this with what I've heard about using the realtime server for messaging services. It would seem that you couldn't make a table public if it contained data about messages sent by users.

**Comment by kiwicopple** [comment]
> any tables with realtime enabled need to be public (for now)?

They don't have to be in the `public` schema - they can exist anywhere in your database. However if you [turn on Realtime](https://supabase.io/docs/guides/api#managing-realtime) for a table it will broadcast changes publicly. This is obviously not ideal for most purposes except if you have tables with "globally accessible" data - things like products, leaderboards, etc.

> couldn't make a table public if it contained data about messages sent by users

If the tables contain private messages, this is correct. We know - very frustrating. The reason it's taking so long is because we want to make sure it scales. The short-term win for Supabase would be just a "polling" strategy, or to re-fetch from the database whenever there is a change (this is what most systems do). We're trying to determine if this can be done without another round-trip to the database (something like [materialize](https://materialize.com/), with Row Level Security). We know it can technically be done, we just need to figure out all the moving parts. Once this is done - we promise that you'll be able to create extremely scalable realtime apps, even with a small database.

**Reply by AndrewHows** [reply]
Is there somewhere we can keep informed on this - an issue, or a git branch or something? This is the big killer for supabase for me atm, especially when comparing with firebase.

**Reply by fabianmossberg** [reply]
Yeah, this is really a feature that I'd love to see!

Generally I have a bigger use for realtime on private tables, such as messages, than I have on any public tables.

**Reply by quicksnap** [reply]
I'd also love to track any tickets!

**Reply by bhvngt** [reply]
Will be great if there is a way to track the progress of this issue. Without RLS, realtime functionality can get very limiting for user authenticated apps

**Reply by sswatson** [reply]
Thanks very much for the explanation. I'm wondering if you have a reference for the recommended way to implement one of the alternative strategies you mentioned.

**Reply by tskytt** [reply]
@kiwicopple Any updates on this? :)

**Reply by inian** [reply]
Quick update - We are working on this here https://github.com/supabase/walrus

**Reply by sswatson** [reply]
Thanks inian! I took a look at that repo, and I have the impression that this is not really intended for developer use yet. E.g., it looks like using that repo would require doing quite a bit of work on the SQL side, whereas one would expect that eventually one could just use realtime features and have the row-level security rules respected. Is that impression accurate?

**Reply by w3b6x9** [reply]
@sswatson I think that's a fair impression.

I created this issue (see: https://github.com/supabase/realtime/issues/177) in Realtime to track [WALRUS](https://github.com/supabase/walrus) integration. We'll post updates there for those who want to track progress!

**Comment by w3b6x9** [comment, marked as answer]
@sswatson just wanted to give you an update that every Supabase project now has Realtime RLS, which means that only authorized users will receive database changes based on your RLS policies.

**Reply by sswatson** [reply]
Thanks for the update. This is great news! 🎉

**Reply by UngererFabien** [reply]
Hello ! That's great ! Does that means [this part of the doc](https://supabase.com/docs/guides/api#details-and-links-2) could be updated where it says: "You should only turn on realtime for Public tables (where all data should be accessible)."

**Reply by w3b6x9** [reply]
@UngererFabien thanks for pointing this out!

Supabase Realtime RLS currently only works for tables under the public schema. I'll update the docs to reflect that.

**Reply by w3b6x9** [reply]
@UngererFabien here's the PR: https://github.com/supabase/supabase/pull/5095
